pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,381 @@
+// Can be used to add additional filters on the outbound path
+func (Plugin) OnOutboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	return nil
+}
+
+// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+// Can be used to add additional filters (e.g., mixer filter) or add more stuff to the HTTP connection manager
+// on the inbound path
+func (Plugin) OnInboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	// Only supports sidecar proxy of HTTP listener for now.
+	if in.Node.Type != model.Sidecar || in.ListenerType != plugin.ListenerTypeHTTP {
+		return nil
+	}
+	if !isRbacEnabled(in.Env.IstioConfigStore) {
+		return nil
+	}
+
+	filter := buildHTTPFilter(in.ServiceInstance.Service.Hostname, in.Env.IstioConfigStore)
+	if filter != nil {
+		for cnum := range mutable.FilterChains {
+			mutable.FilterChains[cnum].HTTP = append(mutable.FilterChains[cnum].HTTP, filter)
+		}
+	}
+
+	return nil
+}
+
+// OnInboundCluster implements the Plugin interface method.
+func (Plugin) OnInboundCluster(env model.Environment, node model.Proxy, service *model.Service,
+	servicePort *model.Port, cluster *xdsapi.Cluster) {
+}
+
+// OnOutboundRouteConfiguration implements the Plugin interface method.
+func (Plugin) OnOutboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.RouteConfiguration) {
+}
+
+// OnInboundRouteConfiguration implements the Plugin interface method.
+func (Plugin) OnInboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.RouteConfiguration) {
+}
+
+// OnOutboundCluster implements the Plugin interface method.
+func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service *model.Service,
+	servicePort *model.Port, cluster *xdsapi.Cluster) {
+}
+
+func isRbacEnabled(store model.IstioConfigStore) bool {
+	var configProto *rbacproto.RbacConfig
+	config := store.RbacConfig(RbacConfigName, kube.IstioNamespace)
+	if config != nil {
+		configProto = config.Spec.(*rbacproto.RbacConfig)
+	}
+	if configProto == nil {
+		log.Debugf("rbac plugin disabled: No RbacConfig")
+		return false
+	}
+	// TODO(yangminzhu): Supports ON_WITH_INCLUSION and ON_WITH_EXCLUSION.
+	if configProto.Mode == rbacproto.RbacConfig_ON {
+		return true
+	}
+
+	log.Debugf("rbac plugin disabled by RbacConfig: %v", *configProto)
+	return false
+}
+
+// buildHTTPFilter builds the RBAC http filter that enforces the access control to the specified
+// service which is co-located with the sidecar proxy.
+func buildHTTPFilter(hostName model.Hostname, store model.IstioConfigStore) *http_conn.HttpFilter {
+	namespace := hostName.Namespace()
+	roles := store.ServiceRoles(namespace)
+	if roles == nil {
+		log.Debugf("no service roles found for %s in namespace %s", hostName, namespace)
+		return nil
+	}
+	bindings := store.ServiceRoleBindings(namespace)
+	config := convertRbacRulesToFilterConfig(hostName.String(), roles, bindings)
```

Comment:
Exit if bindings == nil?

Reply:
Actually this reminded me to not to return nil even when roles == nil. Returning a nil here means no config is to be sent to envoy which means the RBAC filter is not enabled. But I think our logic is deny by default so in this case we should return a default config with empty policies inside.

We don't need to do extra stuff, the convertRbacRulesToFilterConfig() already handle this for us.